### DISCUSSION TITLE
$watch in a component keeps on watching after the component is removed from DOM

### DISCUSSION THREAD

**Discussion by bep** [alpinejs/alpine#2033] [7. Bugs (archived)]
See https://codepen.io/bepsays/pen/bGRgRBG

* If you click any of the blue buttons, the green spans gets incremented by 20, which is correct.
* If you click the red button, the first component is removed from the DOM
* If you then click the blue button I kind of would expect the green counter to be incremented by 10, but it seems that the old watcher is still alive.

Codepen content:

```js
<script>
  document.addEventListener('alpine:init', () => {
    Alpine.store('mystore', {
      counter1: 0,
      counter2: 0
    })
  })
</script>
<div class="mx-auto container">
  <h1>Click 1</h1>
  <div id="c1" x-data="{ }" class="flex space-x-2" x-init="$watch('$store.mystore.counter1', (val) => $store.mystore.counter2 += 10 )">
    <button class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded" @click="$store.mystore.counter1++" x-text="$store.mystore.counter1"></button>
    <span class="bg-green-500 w-20 p-4 font-bold text-white" x-text="$store.mystore.counter2"></span>
  </div>

  <h1>Click 2</h1>
  <div id="c2" x-data="{ }" class="flex space-x-2" x-init="$watch('$store.mystore.counter1', (val) => $store.mystore.counter2 += 10 )">
    <button class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded" @click="$store.mystore.counter1++" x-text="$store.mystore.counter1"></button>
    <span class="bg-green-500 w-20 p-4 font-bold text-white" x-text="$store.mystore.counter2"></span>
    <button class="bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded" @click="document.getElementById('c1').remove()">Remove Component 1</button>
  </div>
</div>
```

**Comment by SimoTod** [comment]
I believe the reason (i realise it could be confused and i don't know if it's desired) is that watch is not really linked to the x-init which created it but it's just linked to the observed data. Once you create the watcher, it will live independently from the dom element.

I assume that is going to be a nightmare with Turbo.

**Comment by bep** [comment]
As a curious side-note, `x-effect` is behaving like I would expect in this situation (the effect stops reacting once the element is removed).

**Reply by SimoTod** [reply]
yep, that follows the implementation. 
Directives are bound to the dom element and reverted when the element is removed from the dom via the cleanup fucntion, magics don't. 
The effect function injected in each directive is, in reality, the returned value of elementBoundEffect (https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives.js#L77) which has its own cleanup callback.
The effect injected into $watch is just a simple one (https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/magics/%24watch.js#L2).

Theoretically, I think it's possible to make $watch bound to the element if @calebporzio is happy with that and would accept a PR.
